### PR TITLE
[Improvement] Optimize quorum write by skipping unnecessary blocks

### PR DESCRIPTION
**Pull request by frankliee**
### What changes were proposed in this pull request?
Write client send blocks in two rounds.
In the primary round, we only send [0, replicaWrite) data.
In the secondary round, we send [replicaWrite, replica) data.
When  the primary round is totally successful, the secondary round can be skipped. 

### Why are the changes needed?
Without this patch, the  quorum writer will send whole replica times of blocks even all RPCs are successful.

### Does this PR introduce _any_ user-facing change?
No.

### How was this patch tested?
All previous UTs and new UTs.

### CODE REVIEW

**Review by colinmjj — APPROVED** [review]
LGTM

**Comment by jerqi** [comment]
Do you add the function switch? And other replicas case.

**Comment by frankliee** [comment]
> Do you add the function switch? And other replicas case.

More cases are adding. 

But is the switch really necessary? Nobody wants to send more data...

**Comment by frankliee** [comment]
> > Do you add the function switch? And other replicas case.
> 
> More cases are adding.
> 
> But is the switch really necessary? Nobody wants to send more data...

And this interface is ready complicated.
ShuffleWriteClientImpl(String clientType, int retryMax, long retryIntervalMax, int heartBeatThreadNum, int replica, int replicaWrite, int replicaRead)

**Comment by jerqi** [comment]
If one server crash, origin implement have better than this.

**Comment by frankliee** [comment]
> If one server crash, origin implement have better than this.

ok

**Comment by frankliee** [comment]
> If one server crash, origin implement have better than this.

This new option will be added after the unified config PR has been merged.

**Comment by jerqi** [comment]
> > If one server crash, origin implement have better than this.
> 
> This new option will be added after the unified config PR has been merged.

OK

**Comment by jerqi** [comment]
LGTM+2

**Review by jerqi — DISMISSED** [review]

File: client-mr/src/main/java/org/apache/hadoop/mapreduce/RssMRConfig.java [hunk 1]
```diff
@@ -50,6 +50,10 @@ public class RssMRConfig {
       MR_RSS_CONFIG_PREFIX + RssClientConfig.RSS_DATA_REPLICA_READ;
   public static final int RSS_DATA_REPLICA_READ_DEFAULT_VALUE =
       RssClientConfig.RSS_DATA_REPLICA_READ_DEFAULT_VALUE;
+  public static final String RSS_DATA_REPLICA_SKIP =
```
Comment (jerqi) [inline client-mr/src/main/java/org/apache/hadoop/mapreduce/RssMRConfig.java:53]: RSS_DATA_REPLICA_SKIP_ENABLED?

File: client/src/main/java/com/tencent/rss/client/impl/ShuffleWriteClientImpl.java [hunk 2]
```diff
@@ -156,41 +172,72 @@ public SendShuffleDataResult sendShuffleData(String appId, List<ShuffleBlockInfo
...
     // todo: better to have fallback solution when send to multiple servers
-    sendShuffleDataAsync(appId, serverToBlocks, serverToBlockIds, successBlockIds, failedBlockIds);
+
+    // sent the primary round of blocks.
+    boolean isAllSuccess = sendShuffleDataAsync(
+        appId, primaryServerToBlocks, primaryServerToBlockIds, blockIdsTracker);
+
+    // The secondary round of blocks is sent only when the primary group issues failed sending.
+    // This should be infrequent.
+    if (!isAllSuccess && replicaSkip) {
```
Comment (jerqi) [inline client/src/main/java/com/tencent/rss/client/impl/ShuffleWriteClientImpl.java:225]: `replicaSkip` is not necessary.

File: client-spark/spark3/src/main/java/org/apache/spark/shuffle/RssShuffleManager.java [hunk 3]
```diff
@@ -78,6 +78,7 @@ public class RssShuffleManager implements ShuffleManager {
   private final int dataReplica;
   private final int dataReplicaWrite;
   private final int dataReplicaRead;
+  private final boolean dataReplicaSkip;
```
Comment (jerqi) [inline client-spark/spark3/src/main/java/org/apache/spark/shuffle/RssShuffleManager.java:81]: enabled.

File: client/src/main/java/com/tencent/rss/client/impl/ShuffleWriteClientImpl.java (hunk 2 above)
Comment (frankliee) [reply client/src/main/java/com/tencent/rss/client/impl/ShuffleWriteClientImpl.java:225]: This is the new option

Comment (jerqi) [reply client/src/main/java/com/tencent/rss/client/impl/ShuffleWriteClientImpl.java:225]: If replicaSkip is enable, collection is empty.We won't send any rpc.It's up to you.

Comment (frankliee) [reply client/src/main/java/com/tencent/rss/client/impl/ShuffleWriteClientImpl.java:225]: Ok, I will use `!primaryServerToBlocks.isEmpty()` instead.

File: client-mr/src/main/java/org/apache/hadoop/mapreduce/RssMRConfig.java (hunk 1 above)
Comment (frankliee) [reply client-mr/src/main/java/org/apache/hadoop/mapreduce/RssMRConfig.java:53]: OK

File: client-spark/spark3/src/main/java/org/apache/spark/shuffle/RssShuffleManager.java (hunk 3 above)
Comment (frankliee) [reply client-spark/spark3/src/main/java/org/apache/spark/shuffle/RssShuffleManager.java:81]: OK

File: client/src/main/java/com/tencent/rss/client/impl/ShuffleWriteClientImpl.java (hunk 2 above)
Comment (jerqi) [reply client/src/main/java/com/tencent/rss/client/impl/ShuffleWriteClientImpl.java:225]: SendaryServerToBlocks.isEmpty()?

Comment (frankliee) [reply client/src/main/java/com/tencent/rss/client/impl/ShuffleWriteClientImpl.java:225]: Yes, copy error...

**Review by jerqi — DISMISSED** [review]

**Review by jerqi — APPROVED** [review]

**Pull request merged**